arcgis-ios-sdk-samples/Display information/Create symbol styles from web styles/CreateSymbolStylesFromWebStylesViewController.swift

```diff
@@ -0,0 +1,295 @@
+    var cachedImages = [AGSSymbol: UIImage]()
+    /// The data source for the legend table.
+    private var symbolsDataSource: SymbolsDataSource?
+    /// A list of symbols created from the web style.
+    private var symbols = [(category: SymbolCategory, symbol: AGSSymbol)]()
+    
+    /// A feature layer with LA County Points of Interest service.
+    let featureLayer: AGSFeatureLayer = {
+        let serviceFeatureTable = AGSServiceFeatureTable(url: URL(string: "http://services.arcgis.com/V6ZHFr6zdgNZuVG0/arcgis/rest/services/LA_County_Points_of_Interest/FeatureServer/0")!)
+        let layer = AGSFeatureLayer(featureTable: serviceFeatureTable)
+        return layer
+    }()
+    
+    // MARK: Methods
+    
+    /// Create a map with an `AGSFeatureLayer` added to its operational layers.
+    /// - Parameter featureLayer: An `AGSFeatureLayer` object.
+    /// - Returns: An `AGSMap` object.
+    func makeMap(featureLayer: AGSFeatureLayer) -> AGSMap {
+        let map = AGSMap(basemapStyle: .arcGISLightGray)
+        map.referenceScale = 1e5
+        map.operationalLayers.add(featureLayer)
+        return map
+    }
+    
+    /// Create an `AGSUniqueValueRenderer` to render feature layer with symbol styles.
+    /// - Parameter fieldNames: The attributes to match the unique values against.
+    /// - Returns: An `AGSUniqueValueRenderer` object.
+    func makeUniqueValueRenderer(fieldNames: [String]) -> AGSUniqueValueRenderer {
+        let renderer = AGSUniqueValueRenderer()
+        renderer.fieldNames = fieldNames
+        renderer.uniqueValues = symbols.flatMap { category, symbol in
+            // For each category value of a symbol, we need to create a
+            // unique value for it, so the field name matches to all categories.
+            category.symbolCategoryValues.map { symbolValue in
+                AGSUniqueValue(description: "", label: category.symbolName, symbol: symbol, values: [symbolValue])
+            }
+        }
+        return renderer
+    }
+    
+    /// Get certain categories of symbols from a symbol style.
+    /// - Parameters:
+    ///   - symbolStyle: An `AGSSymbolStyle` object.
+    ///   - categories: The types of symbols to search in the symbol style.
+    ///   - completion: A closure executed upon success.
+    private func getSymbols(symbolStyle: AGSSymbolStyle, categories: [SymbolCategory], completion: @escaping () -> Void) {
+        let getSymbolsGroup = DispatchGroup()
+        var symbols = [(SymbolCategory, AGSSymbol)]()
+        
+        categories.forEach { category in
+            getSymbolsGroup.enter()
+            symbolStyle.symbol(forKeys: [category.symbolName]) { symbol, _ in
+                defer {
+                    getSymbolsGroup.leave()
+                }
+                // Add the symbol to the result collection and ignore any error.
+                if let symbol = symbol {
+                    symbols.append((category, symbol))
```

Comment:
Instead stopping here and leaving the group, keep going and create the swatch for the symbol. Then once the swatch has been created, create an instance of legend item, append it to the array, and leave the dispatch group. That way there would be no need for an image cache. Is the symbol still needed for the `makeUniqueValueRenderer(fieldNames:)` method? Okay, add it to the legend item.

When refactoring code (like you just did in the last few commits), it is always a good idea to step back and re-evaluate what else can or should be changed. Take the image cache for example. It was only needed so that the table view data source could create an image once and then access it any number of items. The data source wasn't free to modify the array of symbols that was given to it in order to store the images there. But by moving creation of the swatches out of the data source, new possibilities have opened up. Does that make sense?

Reply:
Yeah I totally agree 🤕  Last Friday when I was trying to refactor the code, I felt I'm losing track of some of the earlier suggestions.